stl/inc/barrier

```diff
@@ -0,0 +1,192 @@
+        const ptrdiff_t _Arrival_value = _Arrival._Value & _Barrier_arrival_token_mask;
+        _Arrival._Value                = _Barrier_invalid_token;
+        for (;;) {
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            const ptrdiff_t _Current = _Val._Myval2._Current.load(memory_order_seq_cst);
+            _STL_VERIFY(_Current >= 0, "Invariant counter >= 0, possibly caused by preconditions violation "
+                                       "(N4861 [thread.barrier.class]/12)");
+            if ((_Current & _Barrier_arrival_token_mask) != _Arrival_value) {
+                break;
+            }
+            _Val._Myval2._Current.wait(_Current, memory_order_relaxed);
+        }
+    }
+
+    void arrive_and_wait() noexcept /* strengthened */ {
+        // TRANSITION, GH-1133: should be memory_order_acq_rel
+        ptrdiff_t _Current =
+            _Val._Myval2._Current.fetch_sub(_Barrier_value_step, memory_order_seq_cst) - _Barrier_value_step;
+        const ptrdiff_t _Arrival = _Current & _Barrier_arrival_token_mask;
+        _STL_VERIFY(_Current >= 0, "Precondition: update is less than or equal to the expected count "
+                                   "for the current barrier phase (N4861 [thread.barrier.class]/12)");
+        if ((_Current & _Barrier_value_mask) == 0) {
+            _Completion(_Current);
+            return;
+        }
+
+        for (;;) {
+            _Val._Myval2._Current.wait(_Current, memory_order_relaxed);
+            // TRANSITION, GH-1133: should be memory_order_acquire
+            _Current = _Val._Myval2._Current.load(memory_order_seq_cst);
+            _STL_VERIFY(_Current >= 0, "Invariant counter >= 0, possibly caused by preconditions violation "
+                                       "(N4861 [thread.barrier.class]/12)");
+            if ((_Current & _Barrier_arrival_token_mask) != _Arrival) {
+                break;
+            }
+        }
+    }
+
+    void arrive_and_drop() noexcept /* strengthened */ {
+        const ptrdiff_t _Rem_count =
+            _Val._Myval2._Total.fetch_sub(_Barrier_value_step, memory_order_relaxed) - _Barrier_value_step;
+        _STL_VERIFY(_Rem_count >= 0, "Precondition: The expected count for the current barrier phase "
+                                     "is greater than zero (N4861 [thread.barrier.class]/24) "
+                                     "(checked initial expected count, which is not less than the current)");
+        (void) arrive(1);
+    }
+
+private:
+    void _Completion(const ptrdiff_t _Current) noexcept {
+        const ptrdiff_t _Rem_count = _Val._Myval2._Total.load(memory_order_relaxed);
+        _STL_VERIFY(_Rem_count >= 0, "Invariant: initial expected count less than zero, "
+                                     "possibly caused by preconditions violation "
+                                     "(N4861 [thread.barrier.class]/24)");
+        _Val._Get_first()();

```

Comment:
`is_nothrow_invocable_v` normally means this needs to use `invoke`; however since this is always invoked with 0 arguments it can't be a PMF/PMD and thus a plain `()` is OK. Does that desrve a comment?

Reply:
No, Standard specifies exactly `is_nothrow_invocable` and exactly direct invocation